[5.0.0-rc] [generics]: add `T: ToSchema` bounds on `impl` instead of requiring it on definition?

I have generic types like this:

```rust
#[derive(utoipa::ToSchema)]
struct Signed<T> {
    payload: T,
    sig: String,
}

#[utoipa::path(post, path = "/api", request_body = Signed<String>)]
async fn post_message() {}
```

It fails to compile with:
```
error[E0277]: the trait bound `T: PartialSchema` is not satisfied
 --> src/main.rs:3:14
  |
3 |     payload: T,
  |              ^ the trait `ComposeSchema` is not implemented for `T`, which is required by `T: PartialSchema`
  |
  = note: required for `T` to implement `PartialSchema`
help: consider restricting type parameter `T`
  |
2 | struct Signed<T: utoipa::__dev::ComposeSchema> {
  |                ++++++++++++++++++++++++++++++

error[E0277]: the trait bound `T: ToSchema` is not satisfied
 --> src/main.rs:3:14
  |
3 |     payload: T,
  |              ^ the trait `ToSchema` is not implemented for `T`
  |
help: consider restricting type parameter `T`
  |
2 | struct Signed<T: utoipa::ToSchema> {
  |                ++++++++++++++++++
```

But the suggestion to define `struct Signed<T: utoipa::ToSchema>` would change the public API to require `ToSchema` for any `Signed<T>`. It would break other non-API-related code using `Signed<T>` on non-`ToSchema` types. My use case is to make `utoipa` dependency optional, so it's not possible to introduce breaking change on whether user enables `utoipa` feature.

I think a better solution would be generating the `ToSchema` bound on `impl` block, instead of requiring them on the struct definition. That is:
```rust
#[derive(ToSchema)]
struct Signed<T> { .. }
/// generates:
impl<T: ToSchema /*<- added by derive-macro*/> ToSchema for Signed<T> { .. }
```

--- a/utoipa-gen/CHANGELOG.md
+++ b/utoipa-gen/CHANGELOG.md
@@ -20,6 +20,7 @@
 * Add support for description and summary overriding (https://github.com/juhaku/utoipa/pull/948)
 * Add nest `OpenApi` support (https://github.com/juhaku/utoipa/pull/930)
 * Add support for additional tags via `tags` (https://github.com/juhaku/utoipa/pull/916)
+* Add `bound` attribute for customizing generic impl bounds. (https://github.com/juhaku/utoipa/pull/1079)
 
 ### Fixed
 

--- a/utoipa-gen/src/component/features.rs
+++ b/utoipa-gen/src/component/features.rs
@@ -101,6 +101,7 @@ pub enum Feature {
     ContentEncoding(attributes::ContentEncoding),
     ContentMediaType(attributes::ContentMediaType),
     Discriminator(attributes::Discriminator),
+    Bound(attributes::Bound),
     MultipleOf(validation::MultipleOf),
     Maximum(validation::Maximum),
     Minimum(validation::Minimum),
@@ -212,6 +213,10 @@ impl ToTokensDiagnostics for Feature {
             Feature::ContentEncoding(content_encoding) => quote! { .content_encoding(#content_encoding) },
             Feature::ContentMediaType(content_media_type) => quote! { .content_media_type(#content_media_type) },
             Feature::Discriminator(discriminator) => quote! { .discriminator(Some(#discriminator)) },
+            Feature::Bound(_) => {
+                // specially handled on generating impl blocks.
+                TokenStream::new()
+            }
             Feature::RenameAll(_) => {
                 return Err(Diagnostics::new("RenameAll feature does not support `ToTokens`"))
             }
@@ -294,6 +299,7 @@ impl Display for Feature {
             Feature::ContentEncoding(content_encoding) => content_encoding.fmt(f),
             Feature::ContentMediaType(content_media_type) => content_media_type.fmt(f),
             Feature::Discriminator(discriminator) => discriminator.fmt(f),
+            Feature::Bound(bound) => bound.fmt(f),
         }
     }
 }
@@ -342,6 +348,7 @@ impl Validatable for Feature {
             Feature::ContentEncoding(content_encoding) => content_encoding.is_validatable(),
             Feature::ContentMediaType(content_media_type) => content_media_type.is_validatable(),
             Feature::Discriminator(discriminator) => discriminator.is_validatable(),
+            Feature::Bound(bound) => bound.is_validatable(),
         }
     }
 }
@@ -388,6 +395,7 @@ is_validatable! {
     attributes::ContentEncoding,
     attributes::ContentMediaType,
     attributes::Discriminator,
+    attributes::Bound,
     validation::MultipleOf = true,
     validation::Maximum = true,
     validation::Minimum = true,
@@ -615,6 +623,7 @@ impl_feature_into_inner! {
     attributes::Required,
     attributes::AdditionalProperties,
     attributes::Discriminator,
+    attributes::Bound,
     validation::MultipleOf,
     validation::Maximum,
     validation::Minimum,

--- a/utoipa-gen/src/component/features/attributes.rs
+++ b/utoipa-gen/src/component/features/attributes.rs
@@ -5,7 +5,7 @@ use quote::ToTokens;
 use syn::parse::ParseStream;
 use syn::punctuated::Punctuated;
 use syn::token::Paren;
-use syn::{Error, LitStr, Token, TypePath};
+use syn::{Error, LitStr, Token, TypePath, WherePredicate};
 
 use crate::component::serde::RenameRule;
 use crate::component::{schema, GenericType, TypeTree};
@@ -935,5 +935,34 @@ impl ToTokens for Discriminator {
 impl From<Discriminator> for Feature {
     fn from(value: Discriminator) -> Self {
         Self::Discriminator(value)
+    }
+}
+
+// bound = "GenericTy: Trait"
+impl_feature! {
+    #[derive(Clone)]
+    #[cfg_attr(feature = "debug", derive(Debug))]
+    pub struct Bound(pub(crate) Punctuated<WherePredicate, Token![,]>);
+}
+
+impl Parse for Bound {
+    fn parse(input: syn::parse::ParseStream, _: Ident) -> syn::Result<Self> {
+        let litstr = parse_utils::parse_next(input, || input.parse::<LitStr>())?;
+        let bounds =
+            syn::parse::Parser::parse_str(<Punctuated<_, _>>::parse_terminated, &litstr.value())
+                .map_err(|err| syn::Error::new(litstr.span(), err.to_string()))?;
+        Ok(Self(bounds))
+    }
+}
+
+impl ToTokens for Bound {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        tokens.extend(self.0.to_token_stream())
+    }
+}
+
+impl From<Bound> for Feature {
+    fn from(value: Bound) -> Self {
+        Feature::Bound(value)
     }
 }

--- a/utoipa-gen/src/component/schema.rs
+++ b/utoipa-gen/src/component/schema.rs
@@ -3,8 +3,8 @@ use std::borrow::{Borrow, Cow};
 use proc_macro2::{Ident, TokenStream};
 use quote::{quote, ToTokens};
 use syn::{
-    punctuated::Punctuated, spanned::Spanned, token::Comma, Attribute, Data, Field, Fields,
-    FieldsNamed, FieldsUnnamed, Generics, Path, PathArguments, Variant,
+    parse_quote, punctuated::Punctuated, spanned::Spanned, token::Comma, Attribute, Data, Field,
+    Fields, FieldsNamed, FieldsUnnamed, Generics, Path, PathArguments, Variant,
 };
 
 use crate::{
@@ -24,7 +24,7 @@ use self::{
 
 use super::{
     features::{
-        attributes::{As, Description, RenameAll},
+        attributes::{As, Bound, Description, RenameAll},
         parse_features, pop_feature, Feature, FeaturesExt, IntoInner, ToTokensExt,
     },
     serde::{self, SerdeContainer, SerdeValue},
@@ -70,6 +70,7 @@ impl ToTokensDiagnostics for Schema<'_> {
     fn to_tokens(&self, tokens: &mut TokenStream) -> Result<(), Diagnostics> {
         let ident = self.ident;
         let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
+        let mut where_clause = where_clause.map_or(parse_quote!(where), |w| w.clone());
 
         let root = Root {
             ident,
@@ -112,6 +113,17 @@ impl ToTokensDiagnostics for Schema<'_> {
         } else {
             ident.to_string()
         };
+
+        if let Some(Bound(bound)) = variant.get_schema_bound() {
+            where_clause.predicates.extend(bound.clone());
+        } else {
+            for param in self.generics.type_params() {
+                let param = &param.ident;
+                where_clause
+                    .predicates
+                    .push(parse_quote!(#param : utoipa::ToSchema))
+            }
+        }
 
         tokens.extend(quote! {
             impl #impl_generics utoipa::__dev::ComposeSchema for #ident #ty_generics #where_clause {
@@ -206,6 +218,15 @@ impl<'a> SchemaVariant<'a> {
             _ => [].iter(),
         }
     }
+
+    fn get_schema_bound(&self) -> Option<&Bound> {
+        match self {
+            SchemaVariant::Named(schema) => schema.bound.as_ref(),
+            SchemaVariant::Unnamed(schema) => schema.bound.as_ref(),
+            SchemaVariant::Enum(schema) => schema.bound.as_ref(),
+            SchemaVariant::Unit(_) => None,
+        }
+    }
 }
 
 impl ToTokens for SchemaVariant<'_> {
@@ -235,6 +256,7 @@ pub struct NamedStructSchema {
     tokens: TokenStream,
     pub schema_as: Option<As>,
     fields_references: Vec<SchemaReference>,
+    bound: Option<Bound>,
 }
 
 #[cfg_attr(feature = "debug", derive(Debug))]
@@ -257,6 +279,7 @@ impl NamedStructSchema {
         let schema_as = pop_feature!(features => Feature::As(_) as Option<As>);
         let description: Option<Description> =
             pop_feature!(features => Feature::Description(_)).into_inner();
+        let bound = pop_feature!(features => Feature::Bound(_) as Option<Bound>);
 
         let container_rules = serde::parse_container(root.attributes)?;
 
@@ -448,6 +471,7 @@ impl NamedStructSchema {
             tokens,
             schema_as,
             fields_references,
+            bound,
         })
     }
 
@@ -553,6 +577,7 @@ struct UnnamedStructSchema {
     tokens: TokenStream,
     schema_as: Option<As>,
     schema_references: Vec<SchemaReference>,
+    bound: Option<Bound>,
 }
 
 impl UnnamedStructSchema {
@@ -565,6 +590,7 @@ impl UnnamedStructSchema {
         let schema_as = pop_feature!(features => Feature::As(_) as Option<As>);
         let description: Option<Description> =
             pop_feature!(features => Feature::Description(_)).into_inner();
+        let bound = pop_feature!(features => Feature::Bound(_) as Option<Bound>);
 
         let fields_len = fields.len();
         let first_field = fields.first().unwrap();
@@ -666,6 +692,7 @@ impl UnnamedStructSchema {
             tokens,
             schema_as,
             schema_references,
+            bound,
         })
     }
 }
@@ -681,6 +708,7 @@ pub struct EnumSchema<'a> {
     schema_type: EnumSchemaType<'a>,
     schema_as: Option<As>,
     schema_references: Vec<SchemaReference>,
+    bound: Option<Bound>,
 }
 
 impl<'e> EnumSchema<'e> {
@@ -728,6 +756,8 @@ impl<'e> EnumSchema<'e> {
             };
 
             let schema_as = pop_feature!(features => Feature::As(_) as Option<As>);
+            let bound = pop_feature!(features => Feature::Bound(_) as Option<Bound>);
+
             if parent.attributes.has_deprecated() {
                 features.push(Feature::Deprecated(true.into()))
             }
@@ -736,6 +766,7 @@ impl<'e> EnumSchema<'e> {
                 schema_type: EnumSchemaType::Plain(PlainEnum::new(parent, variants, features)?),
                 schema_as,
                 schema_references: Vec::new(),
+                bound,
             })
         } else {
             let mut enum_features = parent
@@ -744,6 +775,7 @@ impl<'e> EnumSchema<'e> {
                 .into_inner()
                 .unwrap_or_default();
             let schema_as = pop_feature!(enum_features => Feature::As(_) as Option<As>);
+            let bound = pop_feature!(enum_features => Feature::Bound(_) as Option<Bound>);
 
             if parent.attributes.has_deprecated() {
                 enum_features.push(Feature::Deprecated(true.into()))
@@ -754,6 +786,7 @@ impl<'e> EnumSchema<'e> {
                 schema_type: EnumSchemaType::Mixed(mixed_enum),
                 schema_as,
                 schema_references,
+                bound,
             })
         }
     }

--- a/utoipa-gen/src/component/schema/features.rs
+++ b/utoipa-gen/src/component/schema/features.rs
@@ -6,9 +6,9 @@ use syn::{
 use crate::{
     component::features::{
         attributes::{
-            AdditionalProperties, As, ContentEncoding, ContentMediaType, Deprecated, Description,
-            Discriminator, Example, Examples, Format, Inline, Nullable, ReadOnly, Rename,
-            RenameAll, Required, SchemaWith, Title, ValueType, WriteOnly, XmlAttr,
+            AdditionalProperties, As, Bound, ContentEncoding, ContentMediaType, Deprecated,
+            Description, Discriminator, Example, Examples, Format, Inline, Nullable, ReadOnly,
+            Rename, RenameAll, Required, SchemaWith, Title, ValueType, WriteOnly, XmlAttr,
         },
         impl_into_inner, impl_merge, parse_features,
         validation::{
@@ -36,7 +36,8 @@ impl Parse for NamedFieldStructFeatures {
             As,
             crate::component::features::attributes::Default,
             Deprecated,
-            Description
+            Description,
+            Bound
         )))
     }
 }
@@ -57,7 +58,8 @@ impl Parse for UnnamedFieldStructFeatures {
             ValueType,
             As,
             Deprecated,
-            Description
+            Description,
+            Bound
         )))
     }
 }
@@ -76,7 +78,8 @@ impl Parse for EnumFeatures {
             RenameAll,
             As,
             Deprecated,
-            Description
+            Description,
+            Bound
         )))
     }
 }

--- a/utoipa-gen/src/lib.rs
+++ b/utoipa-gen/src/lib.rs
@@ -571,12 +571,12 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 /// # use std::borrow::Cow;
 ///  #[derive(ToSchema)]
 ///  #[schema(as = path::MyType<T>)]
-///  struct Type<T: ToSchema> {
+///  struct Type<T> {
 ///      t: T,
 ///  }
 ///
 ///  #[derive(ToSchema)]
-///  struct Person<'p, T: Sized + ToSchema, P: ToSchema> {
+///  struct Person<'p, T: Sized, P> {
 ///      id: usize,
 ///      name: Option<Cow<'p, str>>,
 ///      field: T,
@@ -585,7 +585,7 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 ///
 ///  #[derive(ToSchema)]
 ///  #[schema(as = path::to::PageList)]
-///  struct Page<T: ToSchema> {
+///  struct Page<T> {
 ///      total: usize,
 ///      page: usize,
 ///      pages: usize,
@@ -594,7 +594,7 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 ///
 ///  #[derive(ToSchema)]
 ///  #[schema(as = path::to::Element<T>)]
-///  enum E<T: ToSchema> {
+///  enum E<T> {
 ///      One(T),
 ///      Many(Vec<T>),
 ///  }

--- a/utoipa-gen/src/lib.rs
+++ b/utoipa-gen/src/lib.rs
@@ -119,6 +119,8 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 ///   OpenAPI spec as _`path.to.Pet`_. This same name will be used throughout the OpenAPI generated
 ///   with `utoipa` when the type is being referenced in [`OpenApi`][openapi_derive] derive macro
 ///   or in [`utoipa::path(...)`][path_macro] macro.
+/// * `bound = ...` Can be used to override default trait bounds on generated `impl`s.
+///   See [Generic schemas section](#generic-schemas) below for more details.
 /// * `default` Can be used to populate default values on all fields using the struct's
 ///   [`Default`] implementation.
 /// * `deprecated` Can be used to mark all fields as deprecated in the generated OpenAPI spec but
@@ -238,6 +240,8 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 ///   OpenAPI spec as _`path.to.Pet`_. This same name will be used throughout the OpenAPI generated
 ///   with `utoipa` when the type is being referenced in [`OpenApi`][openapi_derive] derive macro
 ///   or in [`utoipa::path(...)`][path_macro] macro.
+/// * `bound = ...` Can be used to override default trait bounds on generated `impl`s.
+///   See [Generic schemas section](#generic-schemas) below for more details.
 /// * `deprecated` Can be used to mark the field as deprecated in the generated OpenAPI spec but
 ///   not in the code. If you'd like to mark the field as deprecated in the code as well use
 ///   Rust's own `#[deprecated]` attribute instead.
@@ -265,6 +269,8 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 ///   OpenAPI spec as _`path.to.Pet`_. This same name will be used throughout the OpenAPI generated
 ///   with `utoipa` when the type is being referenced in [`OpenApi`][openapi_derive] derive macro
 ///   or in [`utoipa::path(...)`][path_macro] macro.
+/// * `bound = ...` Can be used to override default trait bounds on generated `impl`s.
+///   See [Generic schemas section](#generic-schemas) below for more details.
 /// * `deprecated` Can be used to mark the enum as deprecated in the generated OpenAPI spec but
 ///   not in the code. If you'd like to mark the enum as deprecated in the code as well use
 ///   Rust's own `#[deprecated]` attribute instead.
@@ -296,6 +302,8 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 ///   OpenAPI spec as _`path.to.Pet`_. This same name will be used throughout the OpenAPI generated
 ///   with `utoipa` when the type is being referenced in [`OpenApi`][openapi_derive] derive macro
 ///   or in [`utoipa::path(...)`][path_macro] macro.
+/// * `bound = ...` Can be used to override default trait bounds on generated `impl`s.
+///   See [Generic schemas section](#generic-schemas) below for more details.
 /// * `deprecated` Can be used to mark the enum as deprecated in the generated OpenAPI spec but
 ///   not in the code. If you'd like to mark the enum as deprecated in the code as well use
 ///   Rust's own `#[deprecated]` attribute instead.
@@ -560,12 +568,15 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 ///
 /// # Generic schemas
 ///
-/// Utoipa supports full set of deeply nested generics as shown below. All the generic types must
-/// implement [`ToSchema`][to_schema] trait and bounds are checked at build time.
+/// Utoipa supports full set of deeply nested generics as shown below. The type will implement
+/// [`ToSchema`][to_schema] if and only if all the generic types implement `ToSchema` by default.
+/// That is in Rust `impl<T> ToSchema for MyType<T> where T: Schema { ... }`.
+/// You can also specify `bound = ...` on the item to override the default auto bounds.
 ///
 /// The _`as = ...`_ attribute is used to define the prefixed or alternative name for the component
 /// in question. This same name will be used throughout the OpenAPI generated with `utoipa` when
 /// the type is being referenced in [`OpenApi`][openapi_derive] derive macro or in [`utoipa::path(...)`][path_macro] macro.
+///
 /// ```rust
 /// # use utoipa::ToSchema;
 /// # use std::borrow::Cow;
@@ -839,6 +850,33 @@ static CONFIG: once_cell::sync::Lazy<utoipa_config::Config> =
 /// ```
 ///
 /// More examples for _`value_type`_ in [`IntoParams` derive docs][into_params].
+///
+/// _**Use `bound` attribute to override the default impl bounds.**_
+///
+/// `bound = ...` accepts a string containing zero or more where-predicates separated by comma, as
+/// the similar syntax to [`serde(bound = ...)`](https://serde.rs/container-attrs.html#bound).
+/// If `bound = ...` exists, the default auto bounds (requiring all generic types to implement
+/// `ToSchema`) will not be applied anymore, and only the specified predicates are added to the
+/// `where` clause of generated `impl` blocks.
+///
+/// ```rust
+/// // Override the default bounds to only require `T: ToSchema`, ignoring unused `U`.
+/// #[derive(utoipa::ToSchema, serde::Serialize)]
+/// #[schema(bound = "T: utoipa::ToSchema")]
+/// struct Partial<T, U> {
+///     used_in_api: T,
+///     #[serde(skip)]
+///     not_in_api: std::marker::PhantomData<U>,
+/// }
+///
+/// // Just remove the auto-bounds. So we got `Unused<T>: ToSchema` for any `T`.
+/// #[derive(utoipa::ToSchema, serde::Serialize)]
+/// #[schema(bound = "")]
+/// struct Unused<T> {
+///     #[serde(skip)]
+///     _marker: std::marker::PhantomData<T>,
+/// }
+/// ```
 ///
 /// [to_schema]: trait.ToSchema.html
 /// [known_format]: openapi/schema/enum.KnownFormat.html

--- a/utoipa-gen/tests/openapi_derive.rs
+++ b/utoipa-gen/tests/openapi_derive.rs
@@ -540,7 +540,7 @@ fn openapi_schemas_resolve_schema_references() {
     use utoipa::ToSchema;
 
     #[derive(ToSchema)]
-    enum Element<T: ToSchema> {
+    enum Element<T> {
         One(T),
         Many(Vec<T>),
     }

--- a/utoipa-gen/tests/openapi_derive_test.rs
+++ b/utoipa-gen/tests/openapi_derive_test.rs
@@ -74,7 +74,7 @@ struct B {
 }
 
 #[derive(Deserialize, Serialize, ToSchema)]
-struct C<T: ToSchema, R: ToSchema> {
+struct C<T, R> {
     field_1: R,
     field_2: T,
 }

--- a/utoipa-gen/tests/path_derive.rs
+++ b/utoipa-gen/tests/path_derive.rs
@@ -2535,7 +2535,7 @@ fn derive_path_test_collect_generic_array_request_body() {
     }
 
     #[derive(ToSchema)]
-    struct CreateRequest<T: ToSchema> {
+    struct CreateRequest<T> {
         value: T,
     }
 
@@ -2624,7 +2624,7 @@ fn derive_path_test_collect_generic_request_body() {
     }
 
     #[derive(ToSchema)]
-    struct CreateRequest<T: ToSchema> {
+    struct CreateRequest<T> {
         value: T,
     }
 

--- a/utoipa-gen/tests/schema_derive_test.rs
+++ b/utoipa-gen/tests/schema_derive_test.rs
@@ -3515,6 +3515,7 @@ fn derive_parse_serde_field_attributes() {
     let post = api_doc! {
         #[derive(Serialize)]
         #[serde(rename_all = "camelCase")]
+        #[schema(bound = "")]
         struct Post<S> {
             #[serde(rename = "uuid")]
             id: String,

--- a/utoipa-gen/tests/schema_generics.rs
+++ b/utoipa-gen/tests/schema_generics.rs
@@ -1,7 +1,29 @@
 use std::borrow::Cow;
+use std::marker::PhantomData;
 
 use utoipa::openapi::{Info, RefOr, Schema};
+use utoipa::openapi::{RefOr, Schema};
+use serde::Serialize;
+use utoipa::openapi::{RefOr, Schema};
 use utoipa::{schema, OpenApi, ToSchema};
+
+#[test]
+fn generic_schema_custom_bound() {
+    #![allow(unused)]
+
+    #[derive(Serialize, ToSchema)]
+    #[schema(bound = "T: Clone + Sized, T: Sized")]
+    struct Type<T> {
+        #[serde(skip)]
+        t: PhantomData<T>,
+    }
+
+    #[derive(Clone)]
+    struct NoToSchema;
+    fn assert_is_to_schema<T: ToSchema>() {}
+
+    assert_is_to_schema::<Type<NoToSchema>>();
+}
 
 #[test]
 fn generic_schema_full_api() {
@@ -9,12 +31,12 @@ fn generic_schema_full_api() {
 
     #[derive(ToSchema)]
     #[schema(as = path::MyType<T>)]
-    struct Type<T: ToSchema> {
+    struct Type<T> {
         t: T,
     }
 
     #[derive(ToSchema)]
-    struct Person<'p, T: Sized + ToSchema, P: ToSchema> {
+    struct Person<'p, T: Sized, P> {
         id: usize,
         name: Option<Cow<'p, str>>,
         field: T,
@@ -23,7 +45,7 @@ fn generic_schema_full_api() {
 
     #[derive(ToSchema)]
     #[schema(as = path::to::PageList)]
-    struct Page<T: ToSchema> {
+    struct Page<T> {
         total: usize,
         page: usize,
         pages: usize,
@@ -32,10 +54,13 @@ fn generic_schema_full_api() {
 
     #[derive(ToSchema)]
     #[schema(as = path::to::Element<T>)]
-    enum E<T: ToSchema> {
+    enum E<T> {
         One(T),
         Many(Vec<T>),
     }
+
+    struct NoToSchema;
+    fn assert_no_need_to_schema_outside_api(_: Type<NoToSchema>) {}
 
     #[utoipa::path(
         get,
@@ -80,12 +105,12 @@ fn schema_macro_run() {
 
     #[derive(ToSchema)]
     #[schema(as = path::MyType<T>)]
-    struct Type<T: ToSchema> {
+    struct Type<T> {
         t: T,
     }
 
     #[derive(ToSchema)]
-    struct Person<'p, T: Sized + ToSchema, P: ToSchema> {
+    struct Person<'p, T: Sized, P> {
         id: usize,
         name: Option<Cow<'p, str>>,
         field: T,
@@ -94,7 +119,7 @@ fn schema_macro_run() {
 
     #[derive(ToSchema)]
     #[schema(as = path::to::PageList)]
-    struct Page<T: ToSchema> {
+    struct Page<T> {
         total: usize,
         page: usize,
         pages: usize,

--- a/utoipa-gen/tests/schema_generics.rs
+++ b/utoipa-gen/tests/schema_generics.rs
@@ -1,10 +1,8 @@
 use std::borrow::Cow;
 use std::marker::PhantomData;
 
-use utoipa::openapi::{Info, RefOr, Schema};
-use utoipa::openapi::{RefOr, Schema};
 use serde::Serialize;
-use utoipa::openapi::{RefOr, Schema};
+use utoipa::openapi::{Info, RefOr, Schema};
 use utoipa::{schema, OpenApi, ToSchema};
 
 #[test]

--- a/utoipa-gen/tests/utoipa_gen_test.rs
+++ b/utoipa-gen/tests/utoipa_gen_test.rs
@@ -112,7 +112,7 @@ struct B {
 }
 
 #[derive(Deserialize, Serialize, ToSchema)]
-struct C<T: ToSchema, R: ToSchema> {
+struct C<T, R> {
     field_1: R,
     field_2: T,
 }

